orttraining/orttraining/test/graph/graph_parser_test.cc

```diff
@@ -0,0 +1,1256 @@
+#include "gtest/gtest.h"
+
+#include "core/common/common.h"
+#include "core/graph/graph.h"
+#include "core/graph/model.h"
+#include "orttraining/core/graph/gradient_builder_base.h"
+#include "orttraining/core/graph/optimizer_builder.h"
+#include "orttraining/core/graph/optimizer_graph_builder.h"
+#include "orttraining/core/graph/allreduce_optimizer_graph_builder.h"
+#if defined(USE_MPI)
+#include "orttraining/core/graph/adasum_optimizer_graph_builder.h"
+#endif
+#include "orttraining/core/graph/zero_optimizer_graph_builder.h"
+#include "test/framework/test_utils.h"
+#include "test/util/include/asserts.h"
+#include "test/test_environment.h"
+#include "orttraining/test/session/training_session_test_utils.h"
+#include "orttraining/core/graph/optimizer_builder.h"
+#include "core/optimizer/utils.h"
+#include "core/graph/contrib_ops/contrib_defs.h"
+
+#include "orttraining/core/graph/graph_parser/pattern_graph.h"
+#include "orttraining/core/graph/graph_parser/matcher.h"
+
+#include <iostream>
+#include <memory>
+
+#define MODEL_FOLDER ORT_TSTR("testdata/transform/")
+
+using onnxruntime::test::CountOpsInGraph;
+using onnxruntime::test::CreateMLValue;
+using onnxruntime::test::TestCPUExecutionProvider;
+using namespace onnxruntime::test::training_session_test_utils;
+
+namespace onnxruntime {
+namespace training {
+namespace test {
+
+using namespace GraphParser;
+
+static void print_graph(const Graph& graph) {
+  GraphViewer graph_viewer(graph);
+  const auto& node_topology_list = graph_viewer.GetNodesInTopologicalOrder();
+  for (auto node_index : node_topology_list) {
+    auto* node = graph.GetNode(node_index);
+    std::cout << "Name: " << node->Name() << ", OpType: " << node->OpType()
+              << ", Domain: " << node->Domain() << ", SinceVersion: " << node->SinceVersion() << std::endl;
+  }
+}
+
+template <typename T>
+bool CheckEmbeddingData(const T* data, int64_t batch_size, int64_t element_count) {
+  // check that all batches has same data.
+  size_t data_length = SafeInt<size_t>(batch_size) * element_count;
+  for (size_t i = gsl::narrow<size_t>(element_count); i < data_length; i++) {
+    if (data[i] != data[i % element_count]) {
+      return false;
+    }
+  }
+  return true;
+}
+
+static NodeArg* ExtractEmbedding(Graph& graph,
```

Comment:
>ExtractEmbedding [](http://example.com/codeflow?start=16&length=16)

are we going to check in this code?

Reply:
I think we can remove this code in PR because it is a part of the replacement phase of ```EmbedLayerFusion``` instead of a part of matching phase. To simplify the code of replacement in unit tests I added it.  :)